pkg/controller/iosvariant/iosvariant_controller.go

```diff
@@ -126,55 +124,51 @@ func (r *ReconcileIOSVariant) Reconcile(request reconcile.Request) (reconcile.Re
 		return reconcile.Result{}, err
 	}
 
-	if foundVariant != "" {
-		// we don't do a full reconciliation (update push app on UPS server based on CR content) but we
-		// only do initial creation of push apps.
-		reqLogger.Info("Skip reconcile: IOS Variant already exists in UPS", "IOSVaraint.Name", instance.Name)
-		return reconcile.Result{}, nil
+	if foundVariant.VariantId == "" {
+
+		createdVariant, err := unifiedpushClient.CreateIOSVariant(instance)
+		if err != nil {
+			reqLogger.Error(err, "Error creating iOS variant in UPS.", "IOSVaraint.Name", instance.Name)
+			return reconcile.Result{RequeueAfter: time.Second * 5}, nil
+		}
+
+		if instance.ObjectMeta.Annotations == nil {
+			instance.ObjectMeta.Annotations = make(map[string]string)
+		}
+		instance.ObjectMeta.Annotations["variantId"] = createdVariant.VariantId
+		err = r.client.Update(context.TODO(), instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		reqLogger.Info("iOS Variant created", "Name", instance.Name, "VariantId", createdVariant.VariantId)
+		return reconcile.Result{Requeue: true}, nil
 	}
 
-	variantId, secret, err := unifiedpushClient.CreateIOSVariant(instance)
-	if err != nil {
-		reqLogger.Error(err, "Error creating iOS variant in UPS.", "IOSVaraint.Name", instance.Name)
-		return reconcile.Result{RequeueAfter: time.Second * 5}, nil
+	if instance.ObjectMeta.Annotations["variantId"] != foundVariant.VariantId {
+		instance.ObjectMeta.Annotations["variantId"] = foundVariant.VariantId
+		err = r.client.Update(context.TODO(), instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
 	}
 
-	instance.Status.VariantId = variantId
-	instance.Status.Secret = secret
-	instance.Status.Ready = true
-	err = r.client.Status().Update(context.TODO(), instance)
-	if err != nil {
-		reqLogger.Error(err, "Error updating IOSVariant status", "IOSVariant.Name", instance.Name)
-		return reconcile.Result{}, err
+	// If either of these have diverged, fix both, so there's just a single status update
+	if instance.Status.VariantId != foundVariant.VariantId || instance.Status.Secret != foundVariant.Secret || instance.Status.Ready != true {
+		instance.Status.VariantId = foundVariant.VariantId
+		instance.Status.Secret = foundVariant.Secret
+		instance.Status.Ready = true
+		err = r.client.Status().Update(context.TODO(), instance)
+		if err != nil {
+			reqLogger.Error(err, "Error updating IOSVariant status", "Name", instance.Name)
+			return reconcile.Result{}, err
+		}
```

Comment:
probably not something we need to fix now, but it does look like there are quite a lot of duplicated code between the android and ios variant controller. Should be able to do something about them.

Reply:
Yeah, I think when I created the controllers originally, I wanted to just have one "Variant" controller, but given that they two different types have mutually exclusive fields, I figured it could save some hassle to just duplicate the shared code for now (rather than trying to determine based on the fields which type it was, and making sure it had all necessary fields but no invalid fields). We can definitely try to share code between them if we find that it's easy and makes sense :+1: